File: Java/Genome_assembly_analysis.java
package projekt;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

import java.util.*;

public class projekt {
    public static void main(String[] args) {
        HashMap<Integer, Set<Integer>> graph = new HashMap<>();
        readData(graph);

        int numComponents = countComponents(graph);
        System.out.println("Number of components with at least three vertices: " + numComponents);

        HashMap<Double, Integer> componentDensities = computeComponentDensity(graph);
        System.out.println(componentDensities);
        

        HashMap<Integer, Integer> degreeDistribution = computeNodeDensity(graph);
        System.out.println(degreeDistribution);
    }

    private static void readData(HashMap<Integer, Set<Integer>> graph) {
        String dataFile1 = "/Users/karin/DA3018/Projekt/Project_DA3018/testrunning/translated_data_1_000_000.txt";
        try (BufferedReader br = new BufferedReader(new FileReader(dataFile1))) {
            System.out.println("Reading data...");
            String line;
            while ((line = br.readLine()) != null) {
                String[] columns = line.trim().split("\\s+");

                int identifier1 = Integer.parseInt(columns[0]);
                int identifier2 = Integer.parseInt(columns[1]);

                // Add edge to the graph
                addEdge(graph, identifier1, identifier2);
            }
            System.out.println("Data read, graph complete");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static void addEdge(HashMap<Integer, Set<Integer>> graph, int vertex1, int vertex2) {
        // Add vertex1 and vertex2 to the graph if they don't exist
        graph.putIfAbsent(vertex1, new HashSet<>());
        graph.putIfAbsent(vertex2, new HashSet<>());

        // Add the edge between vertex1 and vertex2
        graph.get(vertex1).add(vertex2);
        graph.get(vertex2).add(vertex1);
        
        //HASHHashMap FOR NODE, 
    }

    private static int countComponents(HashMap<Integer, Set<Integer>> graph) {
        int count = 0;
        Set<Integer> visited = new HashSet<>();

        for (int vertex : graph.keySet()) {
            if (!visited.contains(vertex)) {
                Set<Integer> component = new HashSet<>();
                dfs(graph, vertex, component, visited);
                if (component.size() >= 3) {
                    count++;
                }
            }
        }

        return count;
    }

    private static void dfs(HashMap<Integer, Set<Integer>> graph, int vertex, Set<Integer> component, Set<Integer> visited) {
        visited.add(vertex);
        component.add(vertex);

        for (int neighbor : graph.getOrDefault(vertex, Collections.emptySet())) {
            if (!visited.contains(neighbor)) {
                dfs(graph, neighbor, component, visited);
            }
        }
    }
    
    private static HashMap<Double, Integer> computeComponentDensity(HashMap<Integer, Set<Integer>> graph) {
    	HashMap<Double, Integer> densityDistribution = new HashMap<>();
    	Set<Integer> visited = new HashSet<>();
    	
    	for (int vertex : graph.keySet())
    		if (!visited.contains(vertex)) {
    			Set<Integer> component = new HashSet<>();
    			dfs(graph, vertex, component, visited);
    			
    			int numNode = component.size();
    			int numEdge = 0;
    			for (int v : component) {
                    numEdge += graph.getOrDefault(v, Collections.emptySet()).size();
                }
    			double density;
    			
    			if (numNode <= 1) {
    				density = 0.0;
    			}
    			
    			else {
                numEdge = numEdge / 2;
                
                double numerator = 2 * numEdge;
                double denominator = numNode * (numNode - 1);
                
                
                density = (double) numerator / denominator;
    			}

                densityDistribution.put(density, densityDistribution.getOrDefault(density, 0) + 1);
            }

        return densityDistribution;
    	
    	
    }
    
    private static HashMap<Integer, Integer> computeNodeDensity(HashMap<Integer, Set<Integer>> graph) {
        HashMap<Integer, Integer> degreeDistribution = new HashMap<>();

        for (Set<Integer> component : graph.values()) {
            int degree = component.size();
            degreeDistribution.put(degree, degreeDistribution.getOrDefault(degree, 0) + 1);
        }

        return degreeDistribution;
    }
    

    

}

File: Python/Genome_assembly_analysis.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed May 31 17:37:50 2023

@author: karin
"""

import networkx as nx
import numpy as np
import matplotlib.pyplot as plt
import time

data_file1 = 'Spruce_fingerprint_2017-03-10_16.48.olp.m4'
data_file2 = 'translated_data.txt'
data_file3 = 'sample_data_1_000_000.txt'

# Create an empty graph
graph = nx.Graph()
def read_data(graph):
    start_time = time.time()
    
    with open(data_file3, 'r') as file:
        print("Reading data...")
        for line in file:
            # Split the line into columns
            columns = line.strip().split()
    
            # Extract relevant information from the columns
            identifier1 = int(columns[0])
            identifier2 = int(columns[1])
            # overlap_start1 = int(columns[2])
            # overlap_end1 = int(columns[3])
            # contig_length1 = int(columns[4])
            # overlap_start2 = int(columns[5])
            # overlap_end2 = int(columns[6])
            # contig_length2 = int(columns[7])

            graph.add_edge(identifier1, identifier2)
        end_time = time.time()
        print("Data read in, graph complete", end_time-start_time, "seconds")


# Count the number of components with at least three vertices
def three_vertices(graph):
    start_time = time.time()
    components = nx.connected_components(graph)
    count = 0
    for component in components:
        if len(component) >= 3:
            count += 1
    end_time = time.time()
    print("3 vertices", end_time-start_time, "seconds")
    return count

def component_density(graph):
    
    start_time = time.time()
        
    components = nx.connected_components(graph)
    densities = []
    for component in components:
        subgraph = graph.subgraph(component)
        num_nodes = subgraph.number_of_nodes()
        num_edges = subgraph.number_of_edges()
        density = 0.0
        if num_nodes > 1:
            max_possible_edges = (num_nodes * (num_nodes - 1)) / 2  # Complete graph
            density = num_edges / max_possible_edges
        densities.append(density)

    # Compute the component density distribution
    density_distribution = {}
    for density in densities:
        if density in density_distribution:
            density_distribution[density] += 1
        else:
            density_distribution[density] = 1
    end_time = time.time()
    print("Component density: ", end_time-start_time, "seconds")
    return density_distribution
        

def node_density(graph):
    start_time = time.time()
    node_degrees = graph.degree()
    
    # Compute the node degree distribution
    degree_distribution = {}
    for node, degree in node_degrees:
        if degree in degree_distribution:
            degree_distribution[degree] += 1
        else:
            degree_distribution[degree] = 1
    end_time = time.time()
    print("Node density: ", end_time-start_time, "seconds")
    return degree_distribution

# Driver code
read_data(graph)

# Compute the number of components with at least three vertices
num_components = three_vertices(graph)
print("Number of components with at least three vertices:", num_components)

# Compute the component density distribution
component_densities = component_density(graph)

# Plot the histogram of component densities
plt.hist(component_densities, bins=10, edgecolor='black')
plt.xlabel("Density")
plt.ylabel("Count")
plt.title("Component Density Distribution")
plt.show()

# Compute the node degree distribution
degree_distribution = node_density(graph)

# Plot the histogram of node degrees
degrees = list(degree_distribution.keys())
counts = list(degree_distribution.values())

plt.hist(degrees, bins=10, edgecolor ='black')
plt.xlabel("Degree")
plt.ylabel("Count")
plt.title("Node Degree Distribution")
  # Set logarithmic scale for the y-axis
plt.show()

